feat: add dynamic file structures in loop using yield-tag

kj-9 opened the pull request:
What I did:
- Added a new Jinja extension in `copier/jinja_ext.py` to handle `yield` tags
  - This is a lightweight extension that detects `yield` tags and sets up the necessary context through `jinja_env`
- Updated `_render_path` to return multiple destination paths and their corresponding contexts when `yield` tags are present
  - Internally, delegated the recursive logic for generating paths and contexts from `yield` tags to the `_render_parts` function
- Added `extra_context` parameter to `_render_file` to enable injection of contexts generated from `yield` tags

kj-9 commented:
hopefully closes #1271

yajo commented:
I only reviewed tests in depth so far, and it seems the interface is almost exactly what I meant! Awesome! 

Quoting from https://github.com/copier-org/copier/issues/1271#issue-1821616806:

> are you rendering a filename?
> No: just let the exception raise and print a nice message "Using yield outside of file names isn't supported".

That seems to be the only missing part. At least I saw no test for that. Could you add this please? 

When done I'll review the implementation. This PR is big!

Thank you very much 😊

kj-9 commented:
@yajo 

Thank you for your review!

I've added a new commit that implements the behavior you mentioned - throwing an exception when yield is used for rendering file contents. I've also added the corresponding test case.

Looking forward to your implementation review!

pawamoy commented on copier/errors.py:
```diff
@@ -112,6 +112,12 @@ def __init__(self, features: Sequence[str]):
 
+class YieldTagInFileError(CopierError):
+    """A yield tag is used in the file content, but it is not allowed."""
+
+    pass
```
`pass` is unnecessary if there's a docstring :slightly_smiling_face: 

```suggestion
```

kj-9 replied:
thanks! I fixed in [6f4acec](https://github.com/copier-org/copier/pull/1855/commits/6f4acecccb67d0f2e896d81ec0580eb499651dbb) .

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+    they are more visible to the final user.
+    """
+
+    yield_context: dict[str, Any]
```
I think that dicts are better suited for variable-length collections, but we know this one will always have two items, so:

```suggestion
    yield_name: str
    yield_iterable: Iterable[Any]
```

This change, of course, implies more changes across the rest of the file, including the docstrings.

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+    as usual. It is caller's responsibility to use the `yield_context` attribute in the
+    template to generate the desired output.
+
+    Example:
```
```suggestion
    !!! example
```

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+    template to generate the desired output.
+
+    Example:
+        template: "{% yield single_var from looped_var %}"
```
The example in the code is self-explanatory. We can remove these lines.
```suggestion
```

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+
+    Example:
+        template: "{% yield single_var from looped_var %}"
+        context: {"looped_var": [1, 2, 3], "single_var": "item"}
```
```suggestion
```

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+        template: "{% yield single_var from looped_var %}"
+        context: {"looped_var": [1, 2, 3], "single_var": "item"}
+
+        then,
```
```suggestion
        ```python
```

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+        >>> template = env.from_string("{% yield single_var from looped_var %}{{ single_var }}{% endyield %}")
+        >>> template.render({"looped_var": [1, 2, 3]})
+        ''
+        >>> env.yield_context
```
Following the suggestion above:

```suggestion
        >>> env.yield_name, env.yield_iterable
```

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,109 @@
+        >>> template.render({"looped_var": [1, 2, 3]})
+        ''
+        >>> env.yield_context
+        {'single_var': 'single_var', 'looped_var': [1, 2, 3]}
```
```suggestion
        ('single_var', [1, 2, 3])
        ```
```

yajo commented on copier/main.py:
```diff
@@ -606,19 +604,29 @@ def _render_template(self) -> None:
+                else:
+                    self._render_file(src_relpath, dst_relpath, extra_context=ctx or {})
+                    if self.jinja_env.yield_context:
+                        raise YieldTagInFileError(
```
I think this raise makes more sense inside `self._render_file` before the render.

yajo commented on copier/main.py:
```diff
@@ -715,8 +727,89 @@ def _render_folder(self, dst_relpath: Path) -> None:
+    def _adjust_rendered_part(self, rendered_part: str) -> str:
+        """Adjust the rendered part if necessary.
+
+        If {{ _copier_conf.answers_file }} becomes the full path,
```
```suggestion
        If `{{ _copier_conf.answers_file }}` becomes the full path,
```

yajo reviewed (changes requested):
Finally, the first code review is here!

Apart from what's mentioned, I also miss the documentation.

Thank you!

kj-9 commented:
@yajo 
Thank you for the quick review!

I have added a few commits:
- Fixes according to your comments
- Minor code refactoring

And in the last commit, I added documentation. Specifically, three points:

1. Added a row for the new feature in the comparison table (as far as I know, only copier supports this feature)
2. Added `jinja_ext.py` to the `Reference`
3. Added a section explaining the new feature to `Configuring`.

This is my first time adding documentation, so I'm not sure if it meets your expectations.
Could you please review it again?

yajo commented on docs/configuring.md:
```diff
@@ -560,6 +560,78 @@ your_project
 You can either use any separator, like `.`, and replace it with `_copier_conf.sep`, like
 in the example above, or just use `/` in the answer (works on Windows too).
 
+## Loop over lists to generate files and directories
```
issue: please move all these docs you're adding to `creating.md`.

yajo commented on docs/configuring.md:
````diff
@@ -1748,3 +1820,7 @@ copier update -a .copier-answers.main.yml
 copier update -a .copier-answers.pre-commit.yml
 copier update -a .copier-answers.ci.yml
 ```
+
+```
+
+```
````
issue: delete this, it seems garbage left by mistake.

kj-9 replied:
deleted in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

yajo reviewed (changes requested):
Implementation looks very good, thanks!

Please review a couple of suggestions for the docs, so we can merge.

@copier-org/maintainers this is an important feature, would you like to add your reviews?

sisp commented:
I'll try to give feedback later today.

pawamoy commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+    """Jinja2 extension for the `yield` tag.
+
+    If `yield` tag is used in a template, this extension sets following attribute to the
+    jinja environment:
```
API docs being rendered through MkDocs, which uses Python-Markdown, bullet lists require a blank line before them.

```suggestion
    jinja environment:

```

kj-9 replied:
done in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

pawamoy commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+    - `yield_name`: The name of the variable that will be yielded.
+    - `yield_iterable`: The variable that will be looped over.
+
+    Note that this extension just sets the attributes but renders template as usual.
```
```suggestion
    Note that this extension just sets the attributes but renders templates as usual.
```

kj-9 replied:
done in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

pawamoy commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+    - `yield_iterable`: The variable that will be looped over.
+
+    Note that this extension just sets the attributes but renders template as usual.
+    It is caller's responsibility to use the `yield_context` attribute in the template to
```
```suggestion
    It is the caller's responsibility to use the `yield_context` attribute in the template to
```

kj-9 replied:
done in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

pawamoy commented on copier/jinja_ext.py:
````diff
@@ -0,0 +1,106 @@
+
+    !!! example
+
+        ```python
````
There's a dedicated language for this: `pycon` (for Python Console).

```suggestion
        ```pycon
```

kj-9 replied:
done in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

pawamoy commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+from jinja2.sandbox import SandboxedEnvironment
+
+
+class YieldEnvironment(SandboxedEnvironment):
```
I wonder if it shouldn't be called `CopierEnvironment`, since it's a custom environment for Copier. Would prevent renaming it later if it is customized further. No strong opinion.

kj-9 replied:
this could be affected by https://github.com/copier-org/copier/pull/1855#discussion_r1899536004
so I leave this for now.

pawamoy commented on copier/main.py:
```diff
@@ -715,8 +728,87 @@ def _render_folder(self, dst_relpath: Path) -> None:
+        yield_name = self.jinja_env.yield_name
+        if yield_name:
+            for value in self.jinja_env.yield_iterable or ():
+                new_context = {**extra_context, **{yield_name: value}}
```
The dict creation + unpacking seems unecessary, though I might be missing something.

```suggestion
                new_context = {**extra_context, yield_name: value}
```

kj-9 replied:
done in https://github.com/copier-org/copier/pull/1855/commits/ac042c5d961790645ba3b0e627c9a6283e64f140

pawamoy commented on copier/main.py:
```diff
@@ -715,8 +728,87 @@ def _render_folder(self, dst_relpath: Path) -> None:
+        if not extra_context:
+            extra_context = {}
+
+        rendered_part = self._render_string(part, extra_context=extra_context)
```
Looks like this could be moved after the `if yield_name` block.

pawamoy replied:
Oh, side-effect, OK :+1:

pawamoy commented on copier/main.py:
```diff
@@ -1114,7 +1188,8 @@ def _apply_update(self) -> None:  # noqa: C901
                         # before storing the file name for marking it as unmerged after the loop.
                         with open(fname) as conflicts_candidate:
                             if any(
-                                line.rstrip() in {"<<<<<<< before updating", ">>>>>>> after updating"}
+                                line.rstrip()
+                                in {"<<<<<<< before updating", ">>>>>>> after updating"}
```
There are a few other style changes like this and I wonder if this is because of a discrepancy again between Copier's config and the contributor's editor/IDE configuration.

kj-9 replied:
This diff was displayed even though it had already been merged into the master branch (https://github.com/copier-org/copier/commit/fd23923b5cc903131fa828f4a89b4f457d1750e6). Please check again after merging the master branch and adding a new commit, as the diff has already disappeared.

pawamoy replied:
Perfect, thank you! :ok_hand:

pawamoy reviewed (commented):
A few suggestions, nothing critical, looking good :slightly_smiling_face:

sisp reviewed (changes requested):
Thanks for working on this long-awaited feature, @kj-9! :pray: I have a few conceptual remarks. WDYT, @kj-9 and @copier-org/maintainers?

yajo commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+from jinja2.sandbox import SandboxedEnvironment
+
+
+class YieldEnvironment(SandboxedEnvironment):
+    """Jinja2 environment with attributes from the YieldExtension.
+
+    This is simple environment class that extends the SandboxedEnvironment
+    for use with the YieldExtension, mainly for avoiding type errors.
+
+    We use the SandboxedEnvironment because we want to minimize the risk of hidden malware
+    in the templates. Of course we still have the post-copy tasks to worry about, but at least
+    they are more visible to the final user.
+    """
+
+    yield_name: str | None
+    yield_iterable: Iterable[Any] | None
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.extend(yield_name=None, yield_iterable=None)
```
> In addition, in light of [factoring out Copier's Jinja extensions](https://github.com/copier-org/copier/pull/1809#issuecomment-2408823031), this extension might get factored out of Copier

I don't think this particular extension has any use outside of copier, so it can be an exception.

kj-9 replied:
@sisp 
Thanks for your comment.

Your opinion sounds good. I do not fully understand the context and impact of factoring the Jinja extension. I feel it is a matter of style or philosophy which can be decided by the core maintainers, but the implementation seems straightforward. (so I'm happy to work on in if I need to👍)

but IMHO, as @yajo commented, if we are not going to extract this extension, there doesn't seem to be a significant difference between the current implementation and your suggestion. The subclass is quite small and doesn't seem to be a significant cost.

yajo replied:
Yes, the yield tag is only to be used in file names, and not within templates, and that's part of the design.

For normal Jinja templates, where all data is within the rendering context, the yield tag doesn't make sense. There's the for tag for that.

So I don't think this extension will be outsourced, and if it were in the future we can always do the extraction later when there is some worldwide need for this extension.

sisp replied:
Agreed, let's keep the environment subclass. :+1:

kj-9 commented on copier/jinja_ext.py:
```diff
@@ -0,0 +1,106 @@
+        Sets the `yield_name` and `yield_iterable` attributes in the environment then calls
+        the provided caller function. If an UndefinedError is raised, it returns an empty string.
+        """
+        self.environment.yield_name = yield_name
+        self.environment.yield_iterable = yield_iterable
```
You are right. I will add treatment for the case and add some test cases as well.

kj-9 replied:
@sisp  @copier-org/maintainers
Added 810f448 to treat multiple yield tags:
- Added an error to be raised when multiple yield tags are detected.
- Added test cases for the error.

Additionally, I found a useful hook called [preprocess](https://jinja.palletsprojects.com/en/stable/extensions/#jinja2.ext.Extension.preprocess), which called every time rendering starts. modified to use this hook to initializes the yield tag variables to `None`.

Please review again🙏
It seems that all other comments have been closed, so this might be the last commit.👍

yajo commented on copier/jinja_ext.py:
````diff
@@ -0,0 +1,126 @@
+
+    !!! example
+
+        ```pycon
````
I didn't know about it, sorry. It's good then.

lbenka commented:
Hi, any idea when this is gonna be released in the main version, seems like it's only on master.

pawamoy commented:
Copier doesn't have a schedule for releases, new releases are made when yajo finds the time to do so, and believes the time is right, mostly :shrug: A better question would be "Are there any blockers I could help with before a new release is published?" :pray: :blush:

yajo commented:
Sorry for the delay, I've been so busy lately...

I'll release when I have time, but you can install from master too in the mean time
